src/backend/distributed/commands/database.c

```diff
@@ -242,3 +262,458 @@ PreprocessAlterDatabaseSetStmt(Node *node, const char *queryString,
 
 	return NodeDDLTaskList(NON_COORDINATOR_NODES, commands);
 }
+
+
+List *
+PreprocessCreateDatabaseStmt(Node *node, const char *queryString,
+							 ProcessUtilityContext processUtilityContext)
+{
+	if (!EnableCreateDatabasePropagation || !ShouldPropagate())
+	{
+		return NIL;
+	}
+
+	EnsureCoordinator();
+
+	/*Validate the statement */
+	DeparseTreeNode(node);
+
+	return NIL;
+}
+
+
+/*
+ * PostprocessCreatedbStmt is executed after the statement is applied to the local
+ * postgres instance. In this stage we can prepare the commands that need to be run on
+ * all workers to create the database.
+ *
+ */
+List *
+PostprocessCreateDatabaseStmt(Node *node, const char *queryString)
+{
+	if (!EnableCreateDatabasePropagation || !ShouldPropagate())
+	{
+		return NIL;
+	}
+
+	EnsureCoordinator();
+
+	char *createDatabaseCommand = DeparseTreeNode(node);
+
+	List *commands = list_make3(DISABLE_DDL_PROPAGATION,
+								(void *) createDatabaseCommand,
+								ENABLE_DDL_PROPAGATION);
+
+	return NontransactionalNodeDDLTask(NON_COORDINATOR_NODES, commands);
+}
+
+
+/*
+ * citus_internal_database_command is an internal UDF to
+ * create/drop a database in an idempotent maner without
+ * transaction block restrictions.
+ */
+Datum
+citus_internal_database_command(PG_FUNCTION_ARGS)
+{
```

Comment:
This UDF needs to be moved into metadata_sync.c with some addtional checks that we perform for citus_internal functions:

```diff
+++	CheckCitusVersion(ERROR);

+++	if (!ShouldSkipMetadataChecks())
+++	{
+++		EnsureCoordinatorInitiatedOperation();
+++	}
+++
+++	PG_ENSURE_ARGNOTNULL(0, "database command");
	text *commandText = PG_GETARG_TEXT_P(0);
	char *command = text_to_cstring(commandText);
	Node *parseTree = ParseTreeNode(command);
```